### ISSUE TITLE
`unused_import`  should not warn on (otherwise used) imports from `alloc` that are in `std` prelude

### ISSUE THREAD

**Issue by jhpratt** [rust-lang/rust#122017]
Demo:

```rust
extern crate alloc;

use alloc::vec::Vec;

fn main() {
    let _ = Vec::<()>::new();
}
```

([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=be5d3d7560af4dfd6e9e7754184dd022))

gives a warning:

```text
warning: the item `Vec` is imported redundantly
   --> src/main.rs:3:5
    |
3   | use alloc::vec::Vec;
    |     ^^^^^^^^^^^^^^^
    |
   ::: /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/prelude/mod.rs:125:13
    |
125 |     pub use super::v1::*;
    |             --------- the item `Vec` is already defined here
    |
    = note: `#[warn(unused_imports)]` on by default
```

Given that imports like this are commonly used when writing code that supports `#![no_std]` via feature flags, I think it would be best to not emit this warning. The exemption, specifically, should be any imports from `alloc` that are in the `std` prelude. It should still warn if never actually used in code.

**Comment by petrochenkov** [comment]
Closing in favor of https://github.com/rust-lang/rust/issues/121708 to centralize the discussion there.
